### ISSUE TITLE
[Payment due date 7/30][$250] Remove "Upload photo" step for users who are adding a group image for the first time

### ISSUE THREAD

**Issue by jamesdeanexpensify** [Expensify/App#42868]
Coming from [this Slack thread](https://expensify.slack.com/archives/C05RECHFBEW/p1717087700449089), the **first time** you go to add a picture to a group chat, we add an extra (unnecessary) click/tap on "Upload photo":

1. You click the camera icon near the group avatar
2. It brings up "Upload photo"
3. You click "Upload photo"
4. It then brings up three options that you can actually do something with (Take photo, Choose from gallery, Choose document)

Recommendation: When it's your **first time** adding a photo to a group chat, skip the "Upload photo" step and go straight to the step with the three options. Video of how it currently looks below. 

https://github.com/Expensify/App/assets/22160301/c941ac65-32ab-4715-a31b-1c48559a9899

<details><summary>Upwork Automation - Do Not Edit</summary>
    <ul>
        <li>Upwork Job URL: https://www.upwork.com/jobs/~01b0f3bb834f330c4f</li>
        <li>Upwork Job ID: 1796274972449415168</li>
        <li>Last Price Increase: 2024-06-06</li>
<li>Automatic offers: </li>
<ul>
<li>b4s36t4 | Contributor | 102899317</li>
</ul></li>
    </ul>
</details>

**Comment by abzokhattab** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.
Remove "Upload photo" step for users who are adding a image for the first time
### What is the root cause of that problem?

if the avatar having the default image then it means its the first time , thus we can enabled the 3 dot menu if the default image is shown 

 
### What changes do you think we should make in order to solve the problem?

[here](https://github.com/Expensify/App/blob/f586090887239001ddef1c4bef3c1440917dcad9/src/components/AvatarWithImagePicker.tsx#L370-L375) we need to add the following condition 

```js
   if (isUsingDefaultAvatar) {
                                    openPicker({
                                        onPicked: showAvatarCropModal,
                                    });
                                    return null;
                                }
```

**Comment by abzokhattab** [comment]
### [Proposal updated](https://github.com/Expensify/App/issues/42868#issuecomment-2140811694)

fixed the rerendering issue

**Comment by b4s36t4** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.
Remove "Upload photo" step for users who are adding a group image for the first time 
### What is the root cause of that problem?

NA - New feature

### What changes do you think we should make in order to solve the problem?

We already have the Attachment Picker setup here https://github.com/Expensify/App/blob/29148f6227940b2121737a1ad448f77429035341/src/components/AvatarWithImagePicker.tsx#L368 which handles showing the both modals where we handle both events i.e `Upload Photo` and three menu options.

We shouldn't distrub the existing implementation as it is wrapped with `AttachmentModal` which by defaults mounted and as well as `AttachmentPicker` is also mounted which can cause issues and can call `openPicker` function everytime we call the `function` even though we return `null`.

The optiomal Solution is to wrap this component https://github.com/Expensify/App/blob/29148f6227940b2121737a1ad448f77429035341/src/components/AvatarWithImagePicker.tsx#L317 with `AttachmentPicker` the same way we did for `AttachmentModal` and handle the logic inside the `Pressable` component.

```js
 <AttachmentPicker type={CONST.ATTACHMENT_PICKER_TYPE.IMAGE}>
                            {({openPicker}) => {
                                return (
                                    <PressableWithoutFeedback
                                    onPress={() => {
                                            if (isUsingDefaultAvatar) {
                                                openPicker({
                                                    onPicked: showAvatarCropModal,
                                                });
                                                return
                                            }
                                            if (disabled && enablePreview && onViewPhotoPress) {
                                                onViewPhotoPress();
                                                return;
                                            }
                                            setIsMenuVisible((prev) => !prev);
                                        }}
                                        ...
                                    />
                                )
                            }}
</AttachmentPicker>

```

This way we leave the existing logic as it is and we de-couple the interactions aside from the `AttachmentModal` to the direct `AvatarWithImagePicker` component itself.


### What alternative solutions did you explore? (Optional)
NA

**Comment by mollfpr** [comment]
Will take a look at this in the morning!

**Comment by mollfpr** [comment]
The ref function solution seems promising, but I think it will be hard to follow where the ref gonna be used. Considering the ref is used outside the `AttachmentPicker`. I'll save this for later.

---

@b4s36t4 I tried your solution, but I got this error when trying to apply it.

![Screenshot 2024-06-07 at 19 15 18](https://github.com/Expensify/App/assets/25520267/c0666610-fce7-4b6f-8f2d-270a972d4cea)

---

@Krishna2323 I got this error `Warning: Error: Too many re-renders. React limits the number of renders to prevent an infinite loop.` on iOS.

https://github.com/Expensify/App/assets/25520267/e2892268-a1e1-4862-aa61-a0215f441640

**Comment by b4s36t4** [comment]
@mollfpr Sorry, I might've missed checking it on web.

The solution for that error is instead of wrapping the `AttachmentPicker` over the `PressableWithoutFeedback` please wrap it over the `Tooltip`. 

Updated code as follows.

<img width="573" alt="Screenshot 2024-06-07 at 6 46 51 PM" src="https://github.com/Expensify/App/assets/59088937/8194b8ee-32d7-4c12-8e4e-067c03077767">

> I'll also update my proposal too.

**Comment by mollfpr** [comment]
@b4s36t4 It's working now. Although, I'm curious if we can just move the `AttachmentPicker` up to wrap both `PressableWithoutFeedback` and `AttachmentModal`, so we don't have to render unnecessary input components and popover while we only use one of them.

**Comment by b4s36t4** [comment]
Yea I also thought the same, but thought it won't be feasible but I'm open to it.

Do you want me to take a look at that approach as well @mollfpr ?

**Comment by mollfpr** [comment]
@b4s36t4 Yeah, let's try it.

**Comment by b4s36t4** [comment]
Sure, I'll check and update you soon.

**Comment by mollfpr** [comment]
@b4s36t4 could you update your proposal with the new implementation?

@jamesdeanexpensify I think we need to add the BZ label here.

**Comment by b4s36t4** [comment]
@mollfpr I did updated can you verify again please?

**Comment by mollfpr** [comment]
@b4s36t4 It seems still wrapping up the `Tooltip` instead all the way to `AttachmentModal`.

**Comment by b4s36t4** [comment]
Ohh sorry, messed it up. Currently I'm outside will update it once I'm with my laptop.

**Comment by b4s36t4** [comment]
@mollfpr I have updated the proposal, can you please verify?!

**Comment by mollfpr** [comment]
The [proposal](https://github.com/Expensify/App/issues/42868#issuecomment-2140899364) from @b4s36t4 looks good to me! It's very convenient to re-placing the component so we can use the function `openPicker` inside the `onPress` from `PressableWithoutFeedback` and keep the current logic for the menu items pop over.

🎀 👀 🎀 C+ reviewed!

**Comment by mollfpr** [comment]
>Do you have any feedback about my alternative solution, where we can use openPicker and show function everywhere in AvatarWithImagePicker component? Thanks.

I feel like it will be over-engineered. If the function `openPicker` and `show` is not passed down to the children, then it would be applicable to use `useImperativeHandle`. So simple as re-arranging the component is more suitable for this case and the current component.

**Comment by mollfpr** [comment]
Thanks for the detail @tienifr!

In summary, your proposal and @b4s36t4 will work well, but I think we want to tackle any performance issue as much as possible. I'll leave the decision to @nkuoch!

**Comment by b4s36t4** [comment]
There are a lot of ways we can improve the **inline function** issue but using `useImperativeHandle` will need to be addressed at all the places we're currently using the component and we need to re-design the component. 

If we're worried that `onItemSelected` being declared as a inline function we could solve this using the following way.

```jsx
const onItemSelected = useCallback(
        (openPicker: any) => (item: any, index: number) => {
            setIsMenuVisible(false);
            // In order for the file picker to open dynamically, the click
            // function must be called from within an event handler that was initiated
            // by the user on Safari.
            if (index === 0 && Browser.isSafari()) {
                openPicker({
                    onPicked: showAvatarCropModal,
                });
            }
        },
        [showAvatarCropModal],
    );
```

and then use the function as a prop without any performance issues and things work as expected without us re-writing the component arch and using `useImperativeHandle`.

**Comment by b4s36t4** [comment]
No it won't, I'm wrapping the function with `useCallback` which will only re-create the function instance only when the dependencies change in which `showAvatarCropModal` is changed. 

We accept the parameters from `onItemSelected` function and return a function which is wrapped with `useCallback` means it's not re-created everytime component re-renders.

**Comment by b4s36t4** [comment]
Imagine you wrap a `useCallback` and think how it works.

```js
const onSelectItem = useCallback(() => () => console.log("hi"))
```

**Comment by b4s36t4** [comment]
Ok, I get your point. I still don't think that's an issue, but to resolve the issue we can use `useMemo` instead of `useCallback`. Why? while `useCallback` memos function `useMemo` memos the value which here the value is function.

```jsx
const onSelectItem = useMemo(() => () => console.log("hi"))
```

**Comment by b4s36t4** [comment]
Although it's impossible to directly pass the `openPicker` to the `useMemo` we can currying concept here to allow passing multiple parameters and multiple times calling the same function without modifying anything.

Implementation as follows

```jsx
const onItemSelected = useMemo(
        () =>
            curry((openPicker: any, item: any, index: number) => {
                //logic here
            }),
        [],
    );
```

We don't need to change anything anywhere else, it just works as it is.

As `useMemo` caches value here `onItemSelected` will always be same on every re-render (except for dependencies).

`curry` function is from `lodash`.

**Comment by Christinadobrzyn** [comment]
Just a heads up - I'm going to be ooo until June 24th. I'm not going to assign anyone new but if you need a new BZ teammate for any reason please feel free to ask for one to be assigned here.

**Comment by mollfpr** [comment]
I have checked our component that uses `useImperativeHandle`. In our practice, we use `useImperativeHandle` when we don't pass down the function we need from the parent component. In this case, we don't have any problem using the function from the parent component because it is passed down to the children.

As I said, it's good to give attention to the performance issue but currently, we don't block any use of the inline function and it is used all over the place. Also, @tienifr if you can show the measurement for the performance issue between the two solutions that would be a huge consideration.

**Comment by mollfpr** [comment]
Friendly bump @nkuoch

**Comment by nkuoch** [comment]
Let's go with @b4s36t4

**Comment by Christinadobrzyn** [comment]
update for Melvin - working on PR https://github.com/Expensify/App/pull/44702

**Comment by Christinadobrzyn** [comment]
update for Melvin - working on PR https://github.com/Expensify/App/pull/44702

**Comment by Christinadobrzyn** [comment]
How's the PR coming @mollfpr and @b4s36t4?

**Comment by mollfpr** [comment]
@Christinadobrzyn we waiting for the fix from @b4s36t4. It not much and I can finish the review after the update.

**Comment by Christinadobrzyn** [comment]
Just checking in on this - The [PR is merged](https://github.com/Expensify/App/pull/44702) but we might have some deploy blockers?

**Comment by b4s36t4** [comment]
No, it's not related to this PR. It is deployed to prod as well, but I'm not sure why Melvin didn't give any message here.

**Comment by b4s36t4** [comment]
@Christinadobrzyn As per this comment https://github.com/Expensify/App/pull/44702#issuecomment-2244526049 this pr did hit the product on the 23rd which I feel is the starting day of the regression period. 

I don't know maybe because of deploy-blocker messages melvin didn't add the review messages here. Can you update the details here so that it won't be missed? 

Thanks!

**Comment by Christinadobrzyn** [comment]
Thanks, @b4s36t4 - I'll add the payment details, as this should be paid today/tomorrow. 

Payouts due:

- [ ] Contributor: $250 @b4s36t4 (can you provide your Upwork profile?)
- [ ] Contributor+: $250 @mollfpr (are you paid through NewDot now?)

Do we need a regression test for this?

**Comment by mollfpr** [comment]
I'm down with the regression test.

1. Open the App, click on the (+) plus icon, and then choose Start Chat
2. Choose multiple people i.e Add to Group
3. Click on Next
4. Click on the avatar or camera icon
5. On Web/Desktop, **Verify** the file explorer is open immediately
6. On Mobile Web, **Verify** it's open the options of the image source or the image explorer
7. On Native, **Verify** it's open the modal with options of image source

**Comment by JmillsExpensify** [comment]
$250 approved for @mollfpr

**Comment by b4s36t4** [comment]
@Christinadobrzyn here's my profile - https://www.upwork.com/freelancers/~018f450a49ceab36aa. I have also accepted the offer sent by melvin when I got assigned to the issue.

**Comment by Christinadobrzyn** [comment]
Regression test - https://github.com/Expensify/Expensify/issues/416898

Paid out based on this payment summary - https://github.com/Expensify/App/issues/42868#issuecomment-2258806140

Closing!
